adapters/mediago/mediago.go

```diff
@@ -0,0 +1,214 @@
+	"github.com/prebid/prebid-server/v2/errortypes"
+	"github.com/prebid/prebid-server/v2/macros"
+	"github.com/prebid/prebid-server/v2/openrtb_ext"
+)
+
+type adapter struct {
+	EndpointTemplate *template.Template
+}
+
+type mediagoResponseBidExt struct {
+	MediaType string `json:"mediaType"`
+}
+
+// Builder builds a new instance of the MediaGo adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	endpoint, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
+	bidder := &adapter{
+		EndpointTemplate: endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var adapterRequests []*adapters.RequestData
+	var errs []error
+
+	adapterRequest, err := a.makeRequest(request)
+	if err == nil {
+		adapterRequests = append(adapterRequests, adapterRequest)
+	} else {
+		errs = append(errs, err)
+	}
+	return adapterRequests, errs
+}
+
+func (a *adapter) makeRequest(request *openrtb2.BidRequest) (*adapters.RequestData, error) {
+
+	mediagoExt, err := getMediaGoExt(request)
+	if err != nil {
+		return nil, err
+	}
+	endPoint, err := a.getEndPoint(mediagoExt)
+	if err != nil {
+		return nil, err
+	}
+
+	preProcess(request)
+	reqBody, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     endPoint,
+		Body:    reqBody,
+		Headers: headers,
+		ImpIDs:  openrtb_ext.GetImpIDs(request.Imp),
+	}, nil
+}
+
+// get MediaGoExt From ext.bidderparams or ext of First Imp. Only check and get first Imp.Ext.Bidder to ExtImpMediago
+func getMediaGoExt(request *openrtb2.BidRequest) (*openrtb_ext.ExtMediaGo, error) {
+	var extMediaGo openrtb_ext.ExtMediaGo
+	var extBidder adapters.ExtImpBidder
+
+	// first get the mediago ext from ext.bidderparams
+	reqExt := &openrtb_ext.ExtRequest{}
+	err := json.Unmarshal(request.Ext, &reqExt)
+	if err != nil {
+		err = json.Unmarshal(reqExt.Prebid.BidderParams, &extMediaGo)
+		if err != nil && extMediaGo.Token != "" {
+			return &extMediaGo, nil
+		}
+	}
+
+	// fallback to get token and region from first imp
+	imp := request.Imp[0]
+	err = json.Unmarshal(imp.Ext, &extBidder)
+	if err != nil {
+		return nil, err
+	}
+
+	var extImpMediaGo openrtb_ext.ExtImpMediaGo
+	err = json.Unmarshal(extBidder.Bidder, &extImpMediaGo)
+	if err != nil {
+		return nil, err
+	}
+	if extImpMediaGo.Token != "" {
+		extMediaGo.Token = extImpMediaGo.Token
+		extMediaGo.Region = extImpMediaGo.Region
+
+		return &extMediaGo, nil
```

Comment:
Based on this code block it seems like adapter code excepts all imps to have same bidder param values.

@SylviaF  could you confirm if this is expected behaviour? If yes then this should be explicitly mentioned in bidder docs

Reply:
@onkarvhanumante looks like @SylviaF made the docs change updates you requested on this pr. 

@SylviaF can you also update the below file in our docs repo to reflect that you are adding a server adapter -> https://github.com/prebid/prebid.github.io/blob/master/dev-docs/bidders/mediago.md

Reply:
> @onkarvhanumante looks like @SylviaF made the docs change updates you requested on this pr.
> 
> @SylviaF can you also update the below file in our docs repo to reflect that you are adding a server adapter -> https://github.com/prebid/prebid.github.io/blob/master/dev-docs/bidders/mediago.md

@SylviaF could you link docs PR in this PR description

Reply:
> Based on this code block it seems like adapter code excepts all imps to have same bidder param values.
> 
> @SylviaF could you confirm if this is expected behaviour? If yes then this should be explicitly mentioned in bidder docs

@onkarvhanumante I'm replying on behalf of @SylviaF : Yes, this is expected behavior. Which file do you mean by bidder doc? Can I submit the doc changes seperately later? I'd like to get the code merged and released in new version in this week.

Reply:
> > @onkarvhanumante looks like @SylviaF made the docs change updates you requested on this pr.
> > @SylviaF can you also update the below file in our docs repo to reflect that you are adding a server adapter -> https://github.com/prebid/prebid.github.io/blob/master/dev-docs/bidders/mediago.md
> 
> @SylviaF could you link docs PR in this PR description

@onkarvhanumante replying on behalf of @SylviaF , she has requested a PR to update the docs: prebid/prebid.github.io#5395